manager/controllers/default/security/login.class.php

```diff
@@ -460,7 +556,95 @@ public function handleForgotLogin() {
+     * Handles the action when a user requests a magic login link.
+     *
+     * @return void
+     * @throws Exception
+     */
+    public function handlePasswordlessLoginRequest() {
+        $this->modx->log(modX::LOG_LEVEL_ERROR, "Handling passwordless login");
+
+        $c = $this->modx->newQuery('modUser');
+        $c->select(['modUser.*', 'Profile.email', 'Profile.fullname']);
+        $c->innerJoin('modUserProfile', 'Profile');
+        $c->where([
+            'Profile.email:=' => $this->scriptProperties['passwordless_login_email'],
+        ]);
+
+        /** @var modUser $user */
+        $user = $this->modx->getObject('modUser', $c);
+
+        if ($user) {
+            $this->modx->log(modX::LOG_LEVEL_DEBUG, "Sending out magic link for user " . $user->get('username'));
+            // we currently just use the forgot password hash
+
+            $activationHash = md5(uniqid(md5($user->get('email') . '/' . $user->get('id')), true));
+
+            $this->setActivationHash($activationHash, $user);
+
+            // Send activation email
+            $message = $this->modx->lexicon('login_magiclink_email');
+            $placeholders = array_merge($this->modx->config, $user->toArray());
+            $placeholders['hash'] = $activationHash;
+            $placeholders['expiration'] = date('D, d M Y H:i',time() + $this->modx->getOption('passwordless_expiration'));
+
+            // Store previous placeholders
+            $ph = $this->modx->placeholders;
+            // now set those useful for modParser
+            $this->modx->setPlaceholders($placeholders);
+            $this->modx->getParser()->processElementTags('', $message, true, false, '[[', ']]', [], 10);
+            $this->modx->getParser()->processElementTags('', $message, true, true, '[[', ']]', [], 10);
+            // Then restore previous placeholders to prevent any breakage
+            $this->modx->placeholders = $ph;
+
+            $this->modx->smarty->assign('config', $this->modx->config);
+            $this->modx->smarty->assign('content', $message);
+            $message = $this->modx->smarty->fetch('email/default.tpl');
+            $sent = $user->sendEmail($message, [
+                'from' => $this->modx->getOption('emailsender'),
+                'fromName' => $this->modx->getOption('site_name'),
+                'sender' => $this->modx->getOption('emailsender'),
+                'subject' => $this->modx->lexicon('login_magiclink_subject'),
+                'html' => true,
+            ]);
+            if (!$sent) {
+                $this->setPlaceholder('error_message', $this->modx->lexicon('error_sending_magiclink_to'));
+            } else {
+                $this->setPlaceholder('success_message', $this->modx->lexicon('login_user_err_nf_email'));
+            }
+        } else {
+            $this->modx->log(modX::LOG_LEVEL_ERROR, "Cannot send magic link, user with email " . $this->scriptProperties['passwordless_login_email'] . " does not exist.");
```

Comment:
Should be removed.

Reply:
I did not remove that log message yet. This can be used to identify brute force attacks in conjunction with e.g. fail2ban easily. Level is now WARN.

Reply:
Fair enough!